File: mytradebot/agent/tools/market.py
"""行情工具：get_quote / get_candles / list_instruments。"""
from __future__ import annotations

from typing import Any

from .registry import ToolDefinition, ToolRegistry, _json_output


def build_market_tools(
    context_provider: Any,
    *,
    candidate_instrument_keys: tuple[str, ...] = tuple(),
) -> ToolRegistry:
    """构建内置行情工具集。context_provider 提供行情数据访问。"""
    registry = ToolRegistry()
    candidate_keys = tuple(key for key in candidate_instrument_keys if key)
    candidate_set = set(candidate_keys)

    def _resolve_key(instrument_key: str) -> str:
        """通过 context_provider 解析标的键名，无解析器时原样返回。"""
        resolver = getattr(context_provider, "resolve_instrument_key", None)
        if callable(resolver):
            return str(resolver(instrument_key))
        return instrument_key

    def _candidate_error(instrument_key: str) -> str | None:
        """检查标的是否在候选范围内，超出则返回错误信息。"""
        if not candidate_set:
            return None
        resolved = _resolve_key(instrument_key)
        if resolved not in candidate_set:
            return f"{resolved} is outside this turn's candidateInstrumentKeys"
        return None

    async def get_quote(instrument_key: str) -> str:
        """获取指定标的的最新报价快照。"""
        candidate_error = _candidate_error(instrument_key)
        if candidate_error:
            return _json_output({"error": candidate_error})
        quote = context_provider.get_quote(instrument_key)
        if quote is None:
            return _json_output({"error": f"No quote available for {instrument_key}"})
        return _json_output({
            "symbol": quote.symbol,
            "price": quote.price,
            "change": quote.change,
            "changePercent": quote.change_percent,
            "dayHigh": quote.day_high,
            "dayLow": quote.day_low,
            "volume": quote.volume,
            "status": quote.status,
        })

    registry.register(ToolDefinition(
        name="get_quote",
        description="获取指定标的的最新报价，包括价格、涨跌幅、成交量等。",
        parameters={
            "type": "object",
            "properties": {
                "instrument_key": {
                    "type": "string",
                    "description": "标的唯一标识，优先使用 list_instruments 返回的 key，如 USDT-FUTURES:BTCUSDT 或 alpaca:AAPL",
                },
            },
            "required": ["instrument_key"],
        },
        handler=get_quote,
    ))

    async def get_candles(
        instrument_key: str,
        count: int = 20,
        interval: str | None = None,
    ) -> str:
        """获取指定标的的最近 K 线数据。"""
        candidate_error = _candidate_error(instrument_key)
        if candidate_error:
            return _json_output({"error": candidate_error})
        candles = context_provider.get_candles(instrument_key, interval=interval)
        if not candles:
            target = f"{instrument_key} @ {interval}" if interval else instrument_key
            return _json_output({"error": f"No candle data for {target}"})
        candle_count = max(1, min(int(count), 50))
        candles = candles[-candle_count:]
        return _json_output([{
            "time": c.open_time_ms // 1000,
            "open": c.open,
            "high": c.high,
            "low": c.low,
            "close": c.close,
            "volume": c.volume,
        } for c in candles])

    registry.register(ToolDefinition(
        name="get_candles",
        description="获取指定标的最近的 OHLCV K 线数据，可指定时间周期和返回数量（最多 50 根）。",
        parameters={
            "type": "object",
            "properties": {
                "instrument_key": {
                    "type": "string",
                    "description": "标的唯一标识",
                },
                "interval": {
                    "type": "string",
                    "description": "可选时间周期，例如 5m、15m、1H、4H、1D；不传时使用当前主周期",
                },
                "count": {
                    "type": "integer",
                    "description": "返回 K 线数量，默认 20，范围 1-50",
                    "default": 20,
                    "minimum": 1,
                    "maximum": 50,
                },
            },
            "required": ["instrument_key"],
        },
        handler=get_candles,
    ))

    async def list_instruments() -> str:
        """列出当前 watchlist 中的所有标的。"""
        instruments = context_provider.list_instruments()
        if candidate_set:
            instruments = tuple(inst for inst in instruments if inst.key in candidate_set)
        return _json_output([{
            "key": inst.key,
            "symbol": inst.symbol,
            "label": inst.label,
            "source": inst.source,
            "group": inst.group,
        } for inst in instruments])

    registry.register(ToolDefinition(
        name="list_instruments",
        description="列出当前 watchlist 中所有激活的标的及其基本信息。",
        parameters={
            "type": "object",
            "properties": {},
        },
        handler=list_instruments,
    ))

    return registry

File: tradex/agent/tools/market.ts
import { QuoteState } from "../../domain/quotes.js";
import { calculateTechnicalIndicators } from "../../domain/indicators.js";
import type { CandleContextMode } from "../../config/index.js";
import { ToolRegistry, jsonOutput } from "./registry.js";
import { shortCandle } from "./market-context.js";

const INTERVAL_ALIASES: Record<string, string> = {
  "1h": "1H",
  "4h": "4H",
  "6h": "6H",
  "12h": "12H",
  "1d": "1D",
  "3d": "3D",
  "1w": "1W",
  "1mo": "1M",
};

export function buildMarketTools(input: { quotes: Record<string, QuoteState>; maxCandles?: number; candleContextMode?: CandleContextMode }): ToolRegistry {
  const registry = new ToolRegistry();
  const resolveKey = (instrumentKey: string) => instrumentKey || Object.keys(input.quotes)[0] || "";
  const maxCandles = Math.max(1, Math.floor(input.maxCandles ?? 80));
  const candleContextMode = input.candleContextMode ?? "raw";

  registry.register({
    name: "get_quote",
    description: "Get the latest quote for an instrument.",
    parameters: { type: "object", properties: { instrument_key: { type: "string" } }, required: ["instrument_key"] },
    execute: ({ instrument_key }) => {
      const key = resolveKey(String(instrument_key || ""));
      const quote = input.quotes[key];
      if (!quote) return jsonOutput({ error: `unknown instrument: ${key}` });
      return jsonOutput({
        instrumentKey: key,
        symbol: quote.symbol,
        displayName: quote.displayName,
        price: quote.price,
        change: quote.change,
        changePercent: quote.changePercent,
        status: quote.status,
        lastError: quote.lastError,
      });
    },
  });

  registry.register({
    name: "get_candles",
    description:
      candleContextMode === "with_indicators"
        ? `Get cached OHLCV candles plus derived RSI/MACD/EMA when enough samples exist. Limit is capped at ${maxCandles}.`
        : `Get cached OHLCV candles for an instrument. Limit is capped at ${maxCandles}.`,
    parameters: {
      type: "object",
      properties: {
        instrument_key: { type: "string" },
        interval: { type: "string", description: "Interval such as 1m, 5m, 15m, 1H, 4H, 1D. Lowercase aliases like 4h are accepted." },
        limit: { type: "integer", minimum: 1, maximum: maxCandles },
      },
      required: ["instrument_key"],
    },
    execute: ({ instrument_key, interval, limit }) => {
      const quote = input.quotes[resolveKey(String(instrument_key || ""))];
      if (!quote) return jsonOutput({ error: "unknown instrument" });
      const intervalKey = resolveIntervalKey(quote.multiTimeframeCandles, interval);
      const candles = intervalKey ? quote.multiTimeframeCandles[intervalKey] ?? [] : quote.candles;
      const requestedLimit = Math.floor(Number(limit));
      const effectiveLimit = Number.isFinite(requestedLimit) && requestedLimit > 0 ? Math.min(requestedLimit, maxCandles) : maxCandles;
      const output: Record<string, unknown> = {
        candles: candles.slice(-effectiveLimit).map(shortCandle),
        interval: intervalKey ?? null,
        limit: effectiveLimit,
      };
      if (candleContextMode === "with_indicators") {
        const indicators = calculateTechnicalIndicators(candles);
        if (indicators !== null) output.indicators = indicators;
      }
      return jsonOutput(output);
    },
  });

  registry.register({
    name: "list_instruments",
    description: "List instruments available to the agent.",
    parameters: { type: "object", properties: {} },
    execute: () => jsonOutput({ instruments: Object.keys(input.quotes) }),
  });

  return registry;
}

function resolveIntervalKey(candlesByInterval: Record<string, unknown>, rawInterval: unknown): string | null {
  if (rawInterval === undefined || rawInterval === null || rawInterval === "") return null;
  const value = String(rawInterval).trim();
  const direct = candlesByInterval[value] !== undefined ? value : null;
  if (direct) return direct;
  const alias = INTERVAL_ALIASES[value] ?? INTERVAL_ALIASES[value.toLowerCase()];
  if (alias && candlesByInterval[alias] !== undefined) return alias;
  return Object.keys(candlesByInterval).find((key) => key.toLowerCase() === value.toLowerCase()) ?? value;
}
